src/main/java/org/jenkinsci/plugins/mesos/api/RunTemplateFactory.java

```diff
@@ -0,0 +1,321 @@
+   * @param containerInfo Optional information for a Docker or Mesos container.
+   * @return the new USI run template.
+   */
+  static RunTemplate newRunTemplate(
+      List<ResourceRequirement> requirements,
+      String shellCommand,
+      String role,
+      List<FetchUri> fetchUris,
+      Optional<MesosAgentSpecTemplate.ContainerInfo> containerInfo) {
+    TaskBuilder taskBuilder =
+        new SimpleTaskInfoBuilder(
+            convertListToSeq(requirements),
+            shellCommand,
+            role,
+            convertListToSeq(fetchUris),
+            Option.empty());
+    if (containerInfo.isPresent()) {
+      taskBuilder = new ContainerInfoTaskInfoBuilder(taskBuilder, containerInfo.get());
+    }
+    return new LegacyLaunchRunTemplate(role, taskBuilder);
+  }
+
+  private static <T> Seq<T> convertListToSeq(List<T> inputList) {
+    return JavaConverters.asScalaIteratorConverter(inputList.iterator()).asScala().toSeq();
+  }
+
+  /**
+   * This is a small USI {@link TaskBuilder} that wraps the {@link SimpleTaskInfoBuilder} and adds
+   * {@link org.apache.mesos.v1.Protos.ContainerInfo} to the Mesos task info if defined.
+   */
+  public static class ContainerInfoTaskInfoBuilder implements TaskBuilder {
+
+    private static final Logger logger =
+        LoggerFactory.getLogger(ContainerInfoTaskInfoBuilder.class);
+
+    public static final String PORT_RESOURCE_NAME = "ports";
+    public static final String MESOS_DEFAULT_ROLE = "*";
+
+    final TaskBuilder simpleTaskInfoBuilder;
+    final MesosAgentSpecTemplate.ContainerInfo containerInfo;
+
+    /**
+     * Constructs a new {@link TaskBuilder}.
+     *
+     * <p>This is basically a port of JenkinsScheduler.getContainerInfoBuilder from v1.1 of the
+     * plugin.
+     *
+     * @param taskInfoBuilder The original {@link SimpleTaskInfoBuilder}.
+     * @param containerInfo The additional container information.
+     */
+    public ContainerInfoTaskInfoBuilder(
+        TaskBuilder taskInfoBuilder, MesosAgentSpecTemplate.ContainerInfo containerInfo) {
+      this.simpleTaskInfoBuilder = taskInfoBuilder;
+      this.containerInfo = containerInfo;
+    }
+
+    @Override
+    public Seq<ResourceRequirement> resourceRequirements() {
+      return this.simpleTaskInfoBuilder.resourceRequirements();
+    }
+
+    @Override
+    public void buildTask(
+        TaskInfo.Builder builder,
+        Offer matchedOffer,
+        Seq<Resource> taskResources,
+        Map<TaskName, Seq<Resource>> peerTaskResources) {
+      this.simpleTaskInfoBuilder.buildTask(builder, matchedOffer, taskResources, peerTaskResources);
+      this.getContainerInfoBuilder(matchedOffer, "what name?", builder);
+    }
+
+    /**
+     * This is the original v1.1 JenkinsScheduler.getContainerInfoBuilder.
+     *
+     * @param offer The Mesos offer.
+     * @param agentName ???
+     * @param taskBuilder The Mesos task info builder.
+     */
+    private void getContainerInfoBuilder(
```

Comment:
This method is almost a verbatim copy of [JenkinsScheduler.getContainerInfoBuilder](https://github.com/jenkinsci/mesos-plugin/blob/1.x/src/main/java/org/jenkinsci/plugins/mesos/JenkinsScheduler.java#L879). I think it should be cleaned up.